compiler/plugins/target/AMD-AIE/iree-amd-aie/Target/XCLBinGen.cpp

```diff
@@ -77,6 +78,107 @@ FailureOr<std::vector<std::string>> flagStringToVector(
                                   std::istream_iterator<std::string>{}};
 }
 
+// Extract an integer from a string, if possible.
+std::optional<int> safeStoi(std::string_view intString) {
+  size_t start = intString.find_first_not_of(" \t\n\r\f\v");
+  if (start == std::string::npos) return std::nullopt;
+  int value = 0;
+  const char *d0 = intString.data() + start;
+  const char *d1 = intString.data() + intString.size();
+  auto [ptr, ec] = std::from_chars(d0, d1, value);
+  if (ec == std::errc()) return value;
+  return std::nullopt;
+}
+
+// We assume that input string is of the form:
+//
+// ```
+// Stack Sizes:
+//      Size     Functions
+//        32     some_func
+//        64     some_other_func
+//       288     core_3_5
+//       288     core_2_5
+//       288     core_1_5
+//       288     core_0_5
+//       288     core_3_4
+//       288     core_3_3
+//       288     core_2_3
+//       288     core_3_2
+//       288     core_1_2
+//       288     core_0_2
+// ```
+//
+// In terms of how we estimate stack sizes, we assume that function call
+// structure is as follows: functions with names core_0_0, core_0_1, core_0_2,
+// et cetera are the entry point functions. These functions call into the
+// other functions like some_func and some_other_func, but never in a
+// nested manner. With these assumptions, an upper bound on the total stack size
+// of a core is the maximum sum of it's stack size, and another function's stack
+// size.
+FailureOr<llvm::DenseMap<std::pair<uint32_t, uint32_t>, uint32_t>>
+getUpperBoundStackSizes(const std::string &readElfOutput) {
+  llvm::DenseMap<std::pair<uint32_t, uint32_t>, uint32_t> coreStackSizes;
+
+  // Split input on whitespace. For the example above, tokens becomes
+  // ['Functions', '32', 'some_func', '64', 'some_other', 288, 'core_3_5', ...]
+  std::vector<std::string> tokens;
```

Comment:
I know this file uses `std::vector` a lot, but `SmallVector` is generally preferred: https://llvm.org/docs/ProgrammersManual.html#llvm-adt-smallvector-h